src/loaders/tiff/lib/indexers.ts

```diff
@@ -0,0 +1,134 @@
+ * images will require extending these indexers or creating new methods.
+ */
+
+/*
+ * Returns an indexer for legacy Bioformats images. This assumes that
+ * downsampled resolutions are stored sequentially in the OME-TIFF.
+ */
+export function getLegacyIndexer(
+  tiff: GeoTIFF,
+  rootMeta: OMEXML
+): OmeTiffIndexer {
+  const imgMeta = rootMeta[0];
+  const { SizeT, SizeC, SizeZ } = imgMeta.Pixels;
+  const ifdIndexer = getIFDIndexer(imgMeta);
+
+  return (sel: OmeTiffSelection, pyramidLevel: number) => {
+    // Get IFD index at base pyramid level
+    const index = ifdIndexer(sel);
+    // Get index of first image at pyramidal level
+    const pyramidIndex = pyramidLevel * SizeZ * SizeT * SizeC;
+    // Return image at IFD index for pyramidal level
+    return tiff.getImage(index + pyramidIndex);
+  };
+}
+
+/*
+ * Returns an indexer for modern Bioforamts images that store multiscale
+ * resolutions using SubIFDs.
+ *
+ * The ifdIndexer returns the 'index' to the base resolution for a
+ * particular 'selection'. The SubIFDs to the downsampled resolutions
+ * of the 'selection' are stored within the `baseImage.fileDirectory`.
+ * We use the SubIFDs to get the IFD for the corresponding sub-resolution.
+ *
+ * NOTE: This function create a custom IFD cache rather than mutating
+ * `GeoTIFF.ifdRequests` with a random offset. The IFDs are cached in
+ * an ES6 Map that maps a string key that identifies the selection uniquely
+ * to the corresponding IFD.
+ */
+export function getSubIFDIndexer(
+  tiff: GeoTIFF,
+  rootMeta: OMEXML
+): OmeTiffIndexer {
+  const imgMeta = rootMeta[0];
+  const ifdIndexer = getIFDIndexer(imgMeta);
+  const ifdCache: Map<string, Promise<ImageFileDirectory>> = new Map();
+
+  return async (sel: OmeTiffSelection, pyramidLevel: number) => {
+    const index = ifdIndexer(sel);
+    const baseImage = await tiff.getImage(index);
+
+    // It's the highest resolution, no need to look up SubIFDs.
+    if (pyramidLevel === 0) {
+      return baseImage;
+    }
+
+    const { SubIFDs } = baseImage.fileDirectory;
+    if (!SubIFDs) {
+      throw Error('Indexing Error: OME-TIFF is missing SubIFDs.');
+    }
+
+    // Get IFD for the selection at the pyramidal level
+    const key = `${sel.t}-${sel.c}-${sel.z}-${pyramidLevel}`;
+    if (!ifdCache.has(key)) {
+      // Only create a new request if we don't have the key.
+      const subIfdOffset = SubIFDs[pyramidLevel - 1];
+      ifdCache.set(key, tiff.parseFileDirectoryAt(subIfdOffset));
+    }
+    const ifd = await ifdCache.get(key) as ImageFileDirectory;
+
+    // Create a new image object manually from IFD
+    // https://github.com/geotiffjs/geotiff.js/blob/8ef472f41b51d18074aece2300b6a8ad91a21ae1/src/geotiff.js#L447-L453
+    return new (baseImage.constructor as any)(
+      ifd.fileDirectory,
+      ifd.geoKeyDirectory,
+      tiff.dataView,
+      tiff.littleEndian,
+      tiff.cache,
+      tiff.source
+    ) as GeoTIFFImage;
+  };
+}
```

Comment:
Probably the most sophisticated addition to this PR.  We keep tract of the key value-mapping to images rather than mutating the tiff.ifdRequests.